---
title: "scGenePanel"
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{scGenePanel}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)

library(dplyr)
library(formattable)
library(htmlwidgets)
library(knitr)
library(kableExtra)
library(stringr)
library(magick)

add_image <- function(filepath, size = 1000){
  image_read(filepath) %>% 
    image_scale(str_extract(size, "[0-9]+"))
}
  
```

Source: `vignettes/scGenePanel.Rmd`

# Introductions

scGenePanel is a tool to visualize single-cell RNA-seq gene expression data. While there are many ways to visualize gene expression, scGenePanel is optimized to export a static image of multiple gene level metrics condensed in one view.

## Features of scGenePanel visual
* A multi-panel plot that contains:
  - UMAP : Visualize cluster of cells, split by cell identity of interest and highlight cell type of interest
  - Violin plot : Visualize gene expression counts, colored by cell identity of interest, split by cell type of interest and significance test computed on cell identity of interest
  - Tabular plot : Quantify cell counts, cell counts >0.25, % expressed and quantiles of gene expression counts per cell identity of interest
* Accepts two commonly used input data object (Seurat, SingleCellExperiment) via object parameter
* "Tableau" or "RColorBrewer" discrete qualitative color palettes options available via col.palette parameter
* Order of cell identity of interest can be changed via group_order parameter
* Shiny version to explore gene expression data interactively

# Installation

From Github repository (once publicly available)

```{r eval=F, echo=T}
library(devtools)
install_github("vandydata/scGenePanel")
```

From local source (Git clone)

```{r eval=F, echo=T}
install.packages("scGenePanel", repos = NULL, type = "source")
#with dependencies
install.packages("scGenePanel", repos = NULL, type = "source", dependencies = TRUE)
```

# Loading data

scGenePanel works natively with any pre-processed Seurat and SingleCellExperiment objects. For this vignette, the dataset used (`human_pancreas-islets_seurat.Rds`) is publicly available at [github](https://github.com/vandydata/CDS-datasets/tree/main/scRNA-Seq). This Seurat object contains single cell RNAseq data of human pancreatic islets from five healthy donors published in Shrestha et al., JCI Insight (2021) that has been <u>downsampled to 500 cells</u>. Key annotations included in the metadata are:
* `CellTypes` - cell types annotations
* `Age` - donor age

If interested, the complete dataset (X cells) is available in [zenodo](https://zenodo.org/record/7626110).

```{r eval=F, echo=T}
library(scGenePanel)
data <- readRDS(url("https://github.com/vandydata/CDS-datasets/raw/main/scRNA-Seq/human_pancreas-islets_seurat.Rds"))
```


# Visualization

`create_gene_panel()` loads the multi-panel visual for the gene of interest among the cell identity of interest entered by parameter `meta_group` and cell type of interest entered by parameter `cell_typ_name`.

```{r eval=F, echo=T}
create_gene_panel(
    object = data,
    gene = "INS",
    meta_group = "Age",
    cell_type_name = "Beta",
    cell_type_colname = "CellTypes",
    col.palette = "Tableau",
    group_order = NULL,
    output_dir = ""
  )
```

```{r, echo = FALSE, fig.retina = 1, dpi = 128}
add_image('/Users/shristishrestha/Documents/GitHub/scGenePanel/genePanel__INS_Beta2.png')
```

# Explore via shiny app

To run the scGenePanel visual via shiny, simply run script below.

```{r eval=F, echo=T}
shiny::runApp("inst/shiny/genepanel")
```


# Session information

```{r}
sessionInfo()
```
